Parsing type from other package fails

The following program does not parse, which should be possible, even if `sub` does not exist.

```go
package main

import (
	"sub"
)

func main() {
	var f sub.Foo
}
```

```
2020/01/03 16:23:50 unable to handle default: 26 - {Type:OPERATOR, Val:., Line:12}
panic: unable to handle default: 26 - {Type:OPERATOR, Val:., Line:12}

goroutine 1 [running]:
log.Panicf(0x11db137, 0x22, 0xc0000e34e0, 0x2, 0x2)
	/Users/gustav/src/go/src/log/log.go:345 +0xc0
github.com/zegl/tre/compiler/parser.(*parser).parseOneWithOptions(0xc0000e3a88, 0x10101, 0xc000092870, 0xc000084840)
	/Users/gustav/src/tre/compiler/parser/parser.go:598 +0x25f5
github.com/zegl/tre/compiler/parser.(*parser).parseOne(0xc0000e3a88, 0x11d4001, 0x1, 0x0)
	/Users/gustav/src/tre/compiler/parser/parser.go:49 +0x3c
github.com/zegl/tre/compiler/parser.(*parser).parseUntilEither(0xc0000e3a88, 0xc0000e3680, 0x1, 0x1, 0x0, 0x3, 0x400, 0xc0000a2320, 0xc0000ac30a, 0xc0000e36a0, ...)
	/Users/gustav/src/tre/compiler/parser/parser.go:975 +0x16f
github.com/zegl/tre/compiler/parser.(*parser).parseUntil(0xc0000e3a88, 0x4, 0x11d40fc, 0x1, 0x0, 0x400, 0x1507600, 0x20300000000000)
	/Users/gustav/src/tre/compiler/parser/parser.go:951 +0x97
github.com/zegl/tre/compiler/parser.(*parser).parseOneWithOptions(0xc0000e3a88, 0x10101, 0x0, 0x0)
	/Users/gustav/src/tre/compiler/parser/parser.go:456 +0xec0
github.com/zegl/tre/compiler/parser.(*parser).parseOne(0xc0000e3a88, 0xc000092801, 0x0, 0x0)
	/Users/gustav/src/tre/compiler/parser/parser.go:49 +0x3c
github.com/zegl/tre/compiler/parser.(*parser).parseUntilEither(0xc0000e3a88, 0xc0000e3a10, 0x1, 0x1, 0xc0000e3a50, 0xe6684204, 0xe7b67620e3df949d, 0x100d648, 0xe700000001197680, 0xc000095001, ...)
	/Users/gustav/src/tre/compiler/parser/parser.go:975 +0x16f
github.com/zegl/tre/compiler/parser.(*parser).parseUntil(0xc0000e3a88, 0xc000095005, 0x0, 0x0, 0x0, 0x18, 0xc0000e3a28, 0xc0000ac3e4)
	/Users/gustav/src/tre/compiler/parser/parser.go:951 +0x97
github.com/zegl/tre/compiler/parser.Parse(0xc0000ea000, 0x35, 0x40, 0x1, 0x40, 0xa2, 0x0)
	/Users/gustav/src/tre/compiler/parser/parser.go:44 +0x2e6
github.com/zegl/tre/cmd/tre/build.parseFile(0x7ffeefbffa39, 0x27, 0x1, 0xc000097a00, 0x0)
	/Users/gustav/src/tre/cmd/tre/build/build.go:183 +0x12f
github.com/zegl/tre/cmd/tre/build.compilePackage(0xc0000c8000, 0x7ffeefbffa39, 0x27, 0xc0000e3f30, 0x2, 0x2, 0x11d43e9, 0x4, 0x19, 0xc000052e68)
	/Users/gustav/src/tre/cmd/tre/build/build.go:105 +0x144
github.com/zegl/tre/cmd/tre/build.Build(0x7ffeefbffa39, 0x27, 0xc0000d3f30, 0x2, 0x2, 0x7ffeefbffa59, 0x4, 0xc000050001, 0x19, 0xc0000ba320)
	/Users/gustav/src/tre/cmd/tre/build/build.go:24 +0xac
main.main()
	/Users/gustav/src/tre/cmd/tre/main.go:52 +0x327
```

--- a/compiler/compiler/alloc.go
+++ b/compiler/compiler/alloc.go
@@ -21,7 +21,7 @@ func (c *Compiler) compileAllocNode(v *parser.AllocNode) {
 	// Allocate from type
 	if len(v.Val) == 1 {
 		if typeNode, ok := v.Val[0].(parser.TypeNode); ok {
-			treType := parserTypeToType(typeNode)
+			treType := c.parserTypeToType(typeNode)
 
 			var alloc *ir.InstAlloca
 
@@ -106,7 +106,7 @@ func (c *Compiler) compileAssignNode(v *parser.AssignNode) {
 	// Allocate from type
 	if typeNode, ok := v.Val[0].(parser.TypeNode); ok {
 		if singleTypeNode, ok := typeNode.(*parser.SingleTypeNode); ok {
-			alloc := c.contextBlock.NewAlloca(parserTypeToType(singleTypeNode).LLVM())
+			alloc := c.contextBlock.NewAlloca(c.parserTypeToType(singleTypeNode).LLVM())
 			dst := c.compileValue(v.Target[0])
 			c.contextBlock.NewStore(c.contextBlock.NewLoad(pointer.ElemType(alloc), alloc), dst.Value)
 			return

--- a/compiler/compiler/array.go
+++ b/compiler/compiler/array.go
@@ -15,7 +15,7 @@ import (
 )
 
 func (c *Compiler) compileInitializeArrayNode(v *parser.InitializeArrayNode) value.Value {
-	itemType := parserTypeToType(v.Type)
+	itemType := c.parserTypeToType(v.Type)
 
 	var values []value.Value
 

--- a/compiler/compiler/compiler.go
+++ b/compiler/compiler/compiler.go
@@ -175,6 +175,19 @@ func (c *Compiler) addGlobal() {
 		IsVariable: false,
 	})
 
+	global.DefinePkgType("bool", types.Bool)
+	global.DefinePkgType("int", types.I64) // TODO: Size based on arch
+	global.DefinePkgType("int8", types.I8)
+	global.DefinePkgType("uint8", types.U8)
+	global.DefinePkgType("int16", types.I16)
+	global.DefinePkgType("uint16", types.U16)
+	global.DefinePkgType("int32", types.I32)
+	global.DefinePkgType("uint32", types.U32)
+	global.DefinePkgType("int64", types.I64)
+	global.DefinePkgType("uint64", types.U64)
+	global.DefinePkgType("uintptr", types.Uintptr)
+	global.DefinePkgType("string", types.String)
+
 	c.packages["global"] = global
 
 	c.module.Funcs = append(c.module.Funcs, strLen)
@@ -208,7 +221,7 @@ func (c *Compiler) compile(instructions []parser.Node) {
 			break
 
 		case *parser.DefineTypeNode:
-			t := parserTypeToType(v.Type)
+			t := c.parserTypeToType(v.Type)
 
 			// Add type to module and override the structtype to use the named
 			// type in the module
@@ -217,7 +230,7 @@ func (c *Compiler) compile(instructions []parser.Node) {
 			}
 
 			// Add to tre mapping
-			typeConvertMap[v.Name] = t
+			c.currentPackage.DefinePkgType(v.Name, t)
 
 		case *parser.SwitchNode:
 			c.compileSwitchNode(v)

--- a/compiler/compiler/condition.go
+++ b/compiler/compiler/condition.go
@@ -68,7 +68,11 @@ func (c *Compiler) compileOperatorNode(v *parser.OperatorNode) value.Value {
 			// Append right to backing array
 			c.contextBlock.NewCall(c.externalFuncs.Strcat.Value.(llvmValue.Named), backingArray, c.contextBlock.NewExtractValue(rightLLVM, 1))
 
-			alloc := c.contextBlock.NewAlloca(typeConvertMap["string"].LLVM())
+			sType, ok := c.packages["global"].GetPkgType("string")
+			if !ok {
+				panic("string type not found")
+			}
+			alloc := c.contextBlock.NewAlloca(sType.LLVM())
 
 			// Save length of the string
 			lenItem := c.contextBlock.NewGetElementPtr(pointer.ElemType(alloc), alloc, constant.NewInt(llvmTypes.I32, 0), constant.NewInt(llvmTypes.I32, 0))
@@ -111,7 +115,7 @@ func (c *Compiler) compileOperatorNode(v *parser.OperatorNode) value.Value {
 		opRes = c.contextBlock.NewXor(leftLLVM, rightLLVM)
 	case parser.OP_BIT_CLEAR:
 		not := c.contextBlock.NewXor(rightLLVM, constant.NewInt(rightLLVM.Type().(*llvmTypes.IntType), -1))
-	 	opRes = c.contextBlock.NewAnd(leftLLVM, not)
+		opRes = c.contextBlock.NewAnd(leftLLVM, not)
 	case parser.OP_LEFT_SHIFT:
 		opRes = c.contextBlock.NewShl(leftLLVM, rightLLVM)
 	case parser.OP_RIGHT_SHIFT:

--- a/compiler/compiler/constants.go
+++ b/compiler/compiler/constants.go
@@ -47,7 +47,11 @@ func (c *Compiler) compileConstantNode(v *parser.ConstantNode) value.Value {
 			c.stringConstants[v.ValueStr] = constString
 		}
 
-		alloc := c.contextBlock.NewAlloca(typeConvertMap["string"].LLVM())
+		sType, ok := c.packages["global"].GetPkgType("string")
+		if !ok {
+			panic("string type not found")
+		}
+		alloc := c.contextBlock.NewAlloca(sType.LLVM())
 
 		// Save length of the string
 		lenItem := c.contextBlock.NewGetElementPtr(pointer.ElemType(alloc), alloc, constant.NewInt(llvmTypes.I32, 0), constant.NewInt(llvmTypes.I32, 0))

--- a/compiler/compiler/func.go
+++ b/compiler/compiler/func.go
@@ -14,12 +14,12 @@ import (
 	"github.com/zegl/tre/compiler/parser"
 )
 
-func funcType(params, returnTypes []parser.TypeNode) (retType types.Type, treReturnTypes []types.Type, argTypes []*ir.Param, treParams []types.Type, isVariadicFunc bool, argumentReturnValuesCount int) {
+func (c *Compiler) funcType(params, returnTypes []parser.TypeNode) (retType types.Type, treReturnTypes []types.Type, argTypes []*ir.Param, treParams []types.Type, isVariadicFunc bool, argumentReturnValuesCount int) {
 	llvmParams := make([]*ir.Param, len(params))
 	treParams = make([]types.Type, len(params))
 
 	for k, par := range params {
-		paramType := parserTypeToType(par)
+		paramType := c.parserTypeToType(par)
 
 		// Variadic arguments are converted into a slice
 		// The function takes a slice as the argument, the caller has to convert
@@ -52,7 +52,7 @@ func funcType(params, returnTypes []parser.TypeNode) (retType types.Type, treRet
 
 	// Use LLVM function return value if there's only one return value
 	if len(returnTypes) == 1 {
-		funcRetType = parserTypeToType(returnTypes[0])
+		funcRetType = c.parserTypeToType(returnTypes[0])
 		treReturnTypes = []types.Type{funcRetType}
 	} else if len(returnTypes) > 0 {
 		// Return values via argument pointers
@@ -60,7 +60,7 @@ func funcType(params, returnTypes []parser.TypeNode) (retType types.Type, treRet
 		var llvmReturnTypesParams []*ir.Param
 
 		for _, ret := range returnTypes {
-			t := parserTypeToType(ret)
+			t := c.parserTypeToType(ret)
 			treReturnTypes = append(treReturnTypes, t)
 			llvmReturnTypesParams = append(llvmReturnTypesParams, ir.NewParam(name.Var("ret"), llvmTypes.NewPointer(t.LLVM())))
 		}
@@ -111,7 +111,7 @@ func (c *Compiler) compileDefineFuncNode(v *parser.DefineFuncNode) value.Value {
 		retTypes[k] = v.Type
 	}
 
-	funcRetType, treReturnTypes, llvmParams, treParams, isVariadicFunc, argumentReturnValuesCount := funcType(argTypes, retTypes)
+	funcRetType, treReturnTypes, llvmParams, treParams, isVariadicFunc, argumentReturnValuesCount := c.funcType(argTypes, retTypes)
 
 	if c.currentPackageName == "main" && v.Name == "main" {
 		if len(v.ReturnValues) != 0 {
@@ -134,7 +134,7 @@ func (c *Compiler) compileDefineFuncNode(v *parser.DefineFuncNode) value.Value {
 
 	// Save as a method on the type
 	if v.IsMethod {
-		if t, ok := typeConvertMap[v.MethodOnType.TypeName]; ok {
+		if t, ok := c.currentPackage.GetPkgType(v.MethodOnType.TypeName); ok {
 			t.AddMethod(v.Name, &types.Method{
 				Function:        typesFunc,
 				LlvmFunction:    fn,

--- a/compiler/compiler/package.go
+++ b/compiler/compiler/package.go
@@ -1,30 +1,39 @@
 package compiler
 
 import (
+	"github.com/zegl/tre/compiler/compiler/types"
 	"github.com/zegl/tre/compiler/compiler/value"
 )
 
 // Representation of a Go package
 type pkg struct {
-	name string
-	vars map[string]value.Value
+	name  string
+	vars  map[string]value.Value
+	types map[string]types.Type
 }
 
 func NewPkg(name string) *pkg {
 	return &pkg{
-		name: name,
+		name:  name,
+		vars:  make(map[string]value.Value),
+		types: make(map[string]types.Type),
 	}
 }
 
 func (p *pkg) DefinePkgVar(name string, val value.Value) {
-	if p.vars == nil {
-		p.vars = make(map[string]value.Value)
-	}
 	p.vars[name] = val
-
 }
 
 func (p *pkg) GetPkgVar(name string) (value.Value, bool) {
 	v, ok := p.vars[name]
+	return v, ok
+}
+
+func (p *pkg) DefinePkgType(name string, ty types.Type) {
+	p.types[name] = ty
+}
+
+func (p *pkg) GetPkgType(name string) (types.Type, bool) {
+	v, ok := p.types[name]
 	return v, ok
 }

--- a/compiler/compiler/slice.go
+++ b/compiler/compiler/slice.go
@@ -56,7 +56,11 @@ func (c *Compiler) compileSubstring(src value.Value, v *parser.SliceArrayNode) v
 	dst := safeBlock.NewCall(c.externalFuncs.Strndup.Value.(llvmValue.Named), offset, length)
 
 	// Convert *i8 to %string
-	alloc := safeBlock.NewAlloca(typeConvertMap["string"].LLVM())
+	sType, ok := c.packages["global"].GetPkgType("string")
+	if !ok {
+		panic("string type not found")
+	}
+	alloc := c.contextBlock.NewAlloca(sType.LLVM())
 
 	// Save length of the string
 	lenItem := safeBlock.NewGetElementPtr(pointer.ElemType(alloc), alloc, constant.NewInt(llvmTypes.I32, 0), constant.NewInt(llvmTypes.I32, 0))
@@ -70,7 +74,7 @@ func (c *Compiler) compileSubstring(src value.Value, v *parser.SliceArrayNode) v
 
 	return value.Value{
 		Value:      safeBlock.NewLoad(pointer.ElemType(alloc), alloc),
-		Type:       typeConvertMap["string"],
+		Type:       sType,
 		IsVariable: false,
 	}
 }
@@ -325,7 +329,7 @@ func (c *Compiler) generateAppendToSliceBlock(appendToSliceBlock *ir.Block, slic
 }
 
 func (c *Compiler) compileInitializeSliceNode(v *parser.InitializeSliceNode) value.Value {
-	itemType := parserTypeToType(v.Type)
+	itemType := c.parserTypeToType(v.Type)
 
 	var values []value.Value
 

--- a/compiler/compiler/struct.go
+++ b/compiler/compiler/struct.go
@@ -110,7 +110,7 @@ func (c *Compiler) compileStructLoadElementNode(v *parser.StructLoadElementNode)
 }
 
 func (c *Compiler) compileInitStructWithValues(v *parser.InitializeStructNode) value.Value {
-	treType := parserTypeToType(v.Type)
+	treType := c.parserTypeToType(v.Type)
 
 	structType, ok := treType.(*types.Struct)
 	if !ok {

--- a/compiler/compiler/types.go
+++ b/compiler/compiler/types.go
@@ -19,21 +19,6 @@ import (
 	llvmValue "github.com/llir/llvm/ir/value"
 )
 
-var typeConvertMap = map[string]types.Type{
-	"bool":    types.Bool,
-	"int":     types.I64, // TODO: Size based on arch
-	"int8":    types.I8,
-	"uint8":   types.U8,
-	"int16":   types.I16,
-	"uint16":  types.U16,
-	"int32":   types.I32,
-	"uint32":  types.U32,
-	"int64":   types.I64,
-	"uint64":  types.U64,
-	"uintptr": types.Uintptr,
-	"string":  types.String,
-}
-
 // Is used in interfaces to keep track of the backing data type
 var typeIDs = map[string]int64{}
 var nextTypeID int64
@@ -48,17 +33,30 @@ func getTypeID(typeName string) int64 {
 	return nextTypeID
 }
 
-func parserTypeToType(typeNode parser.TypeNode) types.Type {
+func (c *Compiler) parserTypeToType(typeNode parser.TypeNode) types.Type {
 	switch t := typeNode.(type) {
 	case *parser.SingleTypeNode:
-		if res, ok := typeConvertMap[t.TypeName]; ok {
+		if len(t.PackageName) > 0 {
+			tp, ok := c.packages[t.PackageName].GetPkgType(t.TypeName)
+			if !ok {
+				panic("unknown type: " + t.PackageName + "." + t.TypeName)
+			}
+			return tp
+		}
+
+		if res, ok := c.currentPackage.GetPkgType(t.TypeName); ok {
+			return res
+		}
+
+		// TODO: Find a better way to organize builtin types
+		if res, ok := c.packages["global"].GetPkgType(t.TypeName); ok {
 			return res
 		}
 
 		panic("unknown type: " + t.TypeName)
 
 	case *parser.ArrayTypeNode:
-		itemType := parserTypeToType(t.ItemType)
+		itemType := c.parserTypeToType(t.ItemType)
 		return &types.Array{
 			Type:     itemType,
 			LlvmType: llvmTypes.NewArray(uint64(t.Len), itemType.LLVM()),
@@ -75,7 +73,7 @@ func parserTypeToType(typeNode parser.TypeNode) types.Type {
 		}
 
 		for i, tt := range t.Types {
-			ty := parserTypeToType(tt)
+			ty := c.parserTypeToType(tt)
 			members[inverseNamesIndex[i]] = ty
 			structTypes = append(structTypes, ty.LLVM())
 		}
@@ -88,7 +86,7 @@ func parserTypeToType(typeNode parser.TypeNode) types.Type {
 		}
 
 	case *parser.SliceTypeNode:
-		itemType := parserTypeToType(t.ItemType)
+		itemType := c.parserTypeToType(t.ItemType)
 		return &types.Slice{
 			Type:     itemType,
 			LlvmType: internal.Slice(itemType.LLVM()),
@@ -104,10 +102,10 @@ func parserTypeToType(typeNode parser.TypeNode) types.Type {
 			}
 
 			for _, arg := range def.ArgumentTypes {
-				ifaceMethod.ArgumentTypes = append(ifaceMethod.ArgumentTypes, parserTypeToType(arg))
+				ifaceMethod.ArgumentTypes = append(ifaceMethod.ArgumentTypes, c.parserTypeToType(arg))
 			}
 			for _, ret := range def.ReturnTypes {
-				ifaceMethod.ReturnTypes = append(ifaceMethod.ReturnTypes, parserTypeToType(ret))
+				ifaceMethod.ReturnTypes = append(ifaceMethod.ReturnTypes, c.parserTypeToType(ret))
 			}
 
 			requiredMethods[name] = ifaceMethod
@@ -117,11 +115,11 @@ func parserTypeToType(typeNode parser.TypeNode) types.Type {
 
 	case *parser.PointerTypeNode:
 		return &types.Pointer{
-			Type: parserTypeToType(t.ValueType),
+			Type: c.parserTypeToType(t.ValueType),
 		}
 
 	case *parser.FuncTypeNode:
-		retType, treReturnTypes, llvmArgTypes, treParams, _, _ := funcType(t.ArgTypes, t.RetTypes)
+		retType, treReturnTypes, llvmArgTypes, treParams, _, _ := c.funcType(t.ArgTypes, t.RetTypes)
 
 		fn := ir.NewFunc("UNNAMEDFUNC", retType.LLVM(), llvmArgTypes...)
 
@@ -147,7 +145,7 @@ func (c *Compiler) compileTypeCastNode(v *parser.TypeCastNode) value.Value {
 		panic("TypeCast origin must be int type")
 	}
 
-	targetType := parserTypeToType(v.Type)
+	targetType := c.parserTypeToType(v.Type)
 	target, ok := targetType.LLVM().(*llvmTypes.IntType)
 	if !ok {
 		panic("TypeCast target must be int type")
@@ -183,7 +181,7 @@ func (c *Compiler) compileTypeCastNode(v *parser.TypeCastNode) value.Value {
 }
 
 func (c *Compiler) compileTypeCastInterfaceNode(v *parser.TypeCastInterfaceNode) value.Value {
-	tryCastToType := parserTypeToType(v.Type)
+	tryCastToType := c.parserTypeToType(v.Type)
 
 	// Allocate the OK variable
 	okVal := c.contextBlock.NewAlloca(types.Bool.LLVM())
@@ -228,8 +226,8 @@ func (c *Compiler) compileTypeCastInterfaceNode(v *parser.TypeCastInterfaceNode)
 			},
 		},
 		MultiValues: []value.Value{
-			value.Value{Type: tryCastToType, Value: resCastedVal, IsVariable: true},
-			value.Value{Type: types.Bool, Value: okVal, IsVariable: true},
+			{Type: tryCastToType, Value: resCastedVal, IsVariable: true},
+			{Type: types.Bool, Value: okVal, IsVariable: true},
 		},
 	}
 }

--- a/compiler/parser/node_types.go
+++ b/compiler/parser/node_types.go
@@ -18,9 +18,10 @@ type TypeNode interface {
 type SingleTypeNode struct {
 	baseNode
 
-	SourceName string
-	TypeName   string
-	IsVariadic bool
+	PackageName string
+	SourceName  string
+	TypeName    string
+	IsVariadic  bool
 }
 
 func (stn SingleTypeNode) Type() string {
@@ -28,7 +29,7 @@ func (stn SingleTypeNode) Type() string {
 }
 
 func (stn SingleTypeNode) String() string {
-	return "type(" + stn.Type() + ")"
+	return fmt.Sprintf("type(%s.%s)", stn.PackageName, stn.Type())
 }
 
 func (stn SingleTypeNode) Variadic() bool {

--- a/compiler/parser/parser.go
+++ b/compiler/parser/parser.go
@@ -1193,6 +1193,20 @@ func (p *parser) parseOneType() (TypeNode, error) {
 	}
 
 	if current.Type == lexer.IDENTIFIER {
+		// Types from other packages ("foo.Bar")
+		isTypeFromPkg := p.lookAhead(1)
+		if isTypeFromPkg.Type == lexer.OPERATOR && isTypeFromPkg.Val == "." {
+			typeName := p.lookAhead(2)
+			p.expect(typeName, lexer.Item{Type: lexer.IDENTIFIER})
+			p.i += 2
+			return &SingleTypeNode{
+				PackageName: current.Val,
+				TypeName:    typeName.Val,
+				IsVariadic:  isVariadic,
+			}, nil
+		}
+
+		// Types from the current package
 		return &SingleTypeNode{
 			TypeName:   current.Val,
 			IsVariadic: isVariadic,

--- a/compiler/testdata/multi-package/main.go
+++ b/compiler/testdata/multi-package/main.go
@@ -1,11 +1,15 @@
-package main  import "external"
+package main
 
-import "sub"
-
-// Hello
-// World
+import (
+	"external"
+	"sub"
+)
 
 func main() {
-	external.Printf("Hello\n")
-	external.Printf("%s\n", sub.World())
+	external.Printf("%s\n", sub.World()) // World
+
+	var a sub.AnotherInt
+	a = 10
+
+	external.Printf("%d\n", a.Plus5()) // 15
 }

